source/language/openpulse.rst

```diff
@@ -0,0 +1,697 @@
+may refer to a transmit channel with an arbitrary name.
+
+.. code-block:: c
+
+    channel d0 = getchannel("drive", $0);  // channel for driving at qubit $0's freq
+    channel cr1_2 = getchannel("coupler", $1, $2);  // channel for a coupler between two qubits
+    channel m2 = getchannel("measure", $2);  // channel for transmitting measurement stimulus
+
+    channel global_field = getchannel("global_field"); // channel not associated with any qubits
+
+    // capture channels for capturing qubits $0 and $1
+    channel cap0 = getchannel("capture", $0);
+    channel cap1 = getchannel("capture", $1);
+
+Frames
+------
+
+When interacting with qubits, it is powerful to track of a frame of reference, akin to the rotating
+frame of a Hamiltonian, throughout the execution of a program. Openpulse provides the ``frame``
+type which is responsible for tracking two properties:
+
+- Tracking time appropriately so programs do not need to deal in absolute time or with the
+  bookkeeping of advancing time in a sequence of pulses.
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
+  defined in the rotating frame with the effect being an equivalent application in the lab
+  frame (i.e. with the carrier supplied by the ``frame``). Another motivation is to more naturally
+  implement a "virtual Z-gate", which does not require a physical pulse but rather shifts the phase
+  of all future pulses on that frame.
+
+The frame is composed of three parts:
+
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
+   is determined by the channel on which the frame is played (see :ref:`Timing` section).
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
+
+Frames can be constructed using the ``newframe`` command e.g.
+
+.. code-block:: javascript
+
+  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)
+
+would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
+instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
+``copyframe`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = newframe(5e9, 0.0);
+  frame driveframe2 = copyframe(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
```

Comment:
Not unique here (we have this elsewhere in OpenQASM, too, but this example involves some kind of implicit casting of `pi/2` to an `angle` type to make this meaningful.

Reply:
Would it be reasonable to always consider `pi` to be an angle?

Reply:
No, but we can address this in the implicit casting rules defined by the types WG.